modules/apps/object/object-service/src/main/java/com/liferay/object/service/impl/ObjectValidationRuleLocalServiceImpl.java

```diff
@@ -478,64 +525,127 @@ private List<ObjectValidationRule> _getObjectValidationRules(
 
 	private List<ObjectValidationRuleSetting>
 			_updateObjectValidationRuleSettings(
-				List<ObjectValidationRuleSetting>
-					newObjectValidationRuleSettings,
-				ObjectValidationRule objectValidationRule)
+				ObjectValidationRule objectValidationRule,
+				List<ObjectValidationRuleSetting> objectValidationRuleSettings)
 		throws PortalException {
 
-		Set<ObjectValidationRuleSetting> objectValidationRuleSettings =
-			SetUtil.fromCollection(newObjectValidationRuleSettings);
+		Map<String, List<String>> objectValidationRuleSettingValuesMap =
+			new HashMap<>();
 
-		for (ObjectValidationRuleSetting oldObjectValidationRuleSetting :
-				_objectValidationRuleSettingPersistence.
-					findByObjectValidationRuleId(
-						objectValidationRule.getObjectValidationRuleId())) {
+		ListUtil.isNotEmptyForEach(
+			objectValidationRuleSettings,
+			objectValidationRuleSetting -> {
+				List<String> objectValidationRuleSettingValues =
+					objectValidationRuleSettingValuesMap.computeIfAbsent(
+						objectValidationRuleSetting.getName(),
+						key -> new ArrayList<>());
 
-			boolean delete = true;
+				objectValidationRuleSettingValues.add(
+					objectValidationRuleSetting.getValue());
+			});
 
-			for (ObjectValidationRuleSetting newObjectValidationRuleSetting :
-					newObjectValidationRuleSettings) {
+		if (StringUtil.equals(
+				objectValidationRule.getEngine(),
+				ObjectValidationRuleConstants.ENGINE_TYPE_COMPOSITE_KEY)) {
 
-				if (StringUtil.equals(
-						oldObjectValidationRuleSetting.getName(),
-						newObjectValidationRuleSetting.getName()) &&
-					StringUtil.equals(
-						oldObjectValidationRuleSetting.getValue(),
-						newObjectValidationRuleSetting.getValue())) {
+			_validateCompositeKeyObjectValidationRuleSettingCount(
+				objectValidationRule.isActive(),
+				objectValidationRuleSettingValuesMap.get(
+					ObjectValidationRuleSettingConstants.
+						NAME_COMPOSITE_KEY_OBJECT_FIELD_ID));
+		}
 
-					delete = false;
+		ObjectDefinition objectDefinition =
+			_objectDefinitionPersistence.fetchByPrimaryKey(
+				objectValidationRule.getObjectDefinitionId());
 
-					objectValidationRuleSettings.remove(
-						newObjectValidationRuleSetting);
+		_deleteObjectValidationRuleSettings(
+			objectValidationRule.getObjectValidationRuleId(),
+			objectDefinition.isApproved(),
+			ObjectValidationRuleSettingConstants.
+				NAME_COMPOSITE_KEY_OBJECT_FIELD_ID,
+			objectValidationRuleSettingValuesMap.get(
+				ObjectValidationRuleSettingConstants.
+					NAME_COMPOSITE_KEY_OBJECT_FIELD_ID));
+		_deleteObjectValidationRuleSettings(
+			objectValidationRule.getObjectValidationRuleId(),
+			objectDefinition.isApproved(),
+			ObjectValidationRuleSettingConstants.NAME_OUTPUT_OBJECT_FIELD_ID,
+			objectValidationRuleSettingValuesMap.get(
+				ObjectValidationRuleSettingConstants.
+					NAME_OUTPUT_OBJECT_FIELD_ID));
```

Comment:
I think this way would be more performatic, besides the fact of call it twice.
Although, I'm doing it separetly because the use of `SetUtil.asymmetricDifference` demands it.